### PR TITLE
Add multi index and disjunctive facet search

### PR DESCRIPTION
**Pull request by bidoubiwa**
⚠️ NOT working with `keepzerofacet`

Using the `Index` widget, it is now possible to do mult-index search, see example bellow.

By introducing the multi index search. we automatically introduce disjunctive facet search.

Related to: #774 #789 

# Multi index search

```jsx
    <InstantSearch indexName="movies" searchClient={searchClient}>
      <SearchBox />
      <h2>Movies</h2>
      <Index indexName="movies">
        <Hits hitComponent={Hit} />
      </Index>
      <h2>Games</h2>
      <Index indexName="games">
        <Hits hitComponent={Hit} />
      </Index>
    </InstantSearch>
```
https://user-images.githubusercontent.com/33010418/203822998-8fc99a62-970f-42a9-95bb-ab09d3c67a9d.mp4

## Disjunctive facet search

https://user-images.githubusercontent.com/33010418/204564207-576bca0c-344b-427c-a1a1-e10ddd7a677f.mp4

# Multi index + disjunctive facet search

https://user-images.githubusercontent.com/33010418/204565010-d4564fdb-88a0-46ad-8dc0-bb6761798e49.mp4

## TODO

- [x] Should work on simple search
- [x] Should be compatible with following widgets:
    - [x] Pagination
    - [x] RefinementList
- [x] Should be compatible with all Meilisearch parameters
    - [x] placeholder search
    - [x] finite pagination
    - [ ] keep zero facets ⚠️ Fixed in disjunctive facet search PR
- [x] check state of cache

### CODE REVIEW

File: src/client/instant-meilisearch-client.ts [hunk 1]
```diff
@@ -67,40 +67,48 @@ export function instantMeiliSearch(
...
           )
-          searchContext.defaultFacetDistribution = defaultFacetDistribution
-        }
 
-        // Search response from Meilisearch
-        const searchResponse = await searchResolver.searchResponse(
-          searchContext,
-          adaptedSearchRequest
-        )
+          // Adapt the Meilisearch responsne to a compliant instantsearch.js response
```
Comment (mdubus) [inline src/client/instant-meilisearch-client.ts:102]:
```suggestion
          // Adapt the Meilisearch response to a compliant instantsearch.js response
```

File: tsconfig.test.json [hunk 2]
```diff
@@ -2,6 +2,7 @@
   "extends": "./tsconfig.json",
   "compilerOptions": {
     "suppressImplicitAnyIndexErrors": true,
+    "resolveJsonModule": true
```
Comment (mdubus) [inline tsconfig.test.json:5]: Out of curiosity, what does this do? 👀

Comment (bidoubiwa) [reply tsconfig.test.json:5]: It allows importing `json` files in typescript files. You cannot fo this for example `import x from 'blabla.json'` by default.

**Review by mdubus — APPROVED** [review]
LGTM ✨ 🦕

File: README.md [hunk 3]
```diff
@@ -358,7 +358,7 @@ The `searchBox` widget is used to let the user perform a text-based query.
 - ✅ container: The CSS Selector or HTMLElement to insert the widget into. _required_
 - ✅ placeholder: Placeholder of the search box.
 - ✅ autofocus: Whether the search box is focused on arrival.
-- ✅ searchAsYouType: Whether result appears as you type or after pressing enter.
+- ✅ searchAsYouType: Whether result appears as you type oxr after pressing enter.
```
Comment (mdubus) [inline README.md:359]: 👀

File: tests/env/react/src/App.js [hunk 4]
```diff
@@ -1,100 +1,19 @@
...
-      <div className="right-panel">
-        <SearchBox />
-        <InfiniteHits hitComponent={Hit} />
-      </div>
-    </InstantSearch>
-  </div>
-)
-
-const Hit = ({ hit }) => {
+// Issue on two different refinment list
```
Comment (mdubus) [inline tests/env/react/src/App.js:8]:
```suggestion
// Issue on two different refinement list
```

File: tests/env/react/src/components/MultiIndex.js [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        <Highlight attribute="name" hit={hit} />
+      </div>
+      <div className="hit-name">
+        <Highlight attribute="title" hit={hit} />
+      </div>
+    </div>
+  )
+}
+
+// Issue on two different refinment list
```
Comment (mdubus) [inline tests/env/react/src/components/MultiIndex.js:32]:
```suggestion
// Issue on two different refinement list
```

File: tests/multi-index-search.tests.ts [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+          facetFilters: [['genres:Adventure', 'genres:Action'], ['color:blue']],
+          facets: ['genres', 'color', 'platforms'],
+        },
+      },
+      {
+        indexName: 'movies',
+        params: {
+          page: 0,
+          hitsPerPage: 1,
+          // @ts-ignore
```
Comment (mdubus) [inline tests/multi-index-search.tests.ts:147]: Out of curiosity, what are you ignoring here?

File: README.md [hunk 7]
```diff
@@ -339,15 +339,13 @@ const search = instantsearch({
...
+### ✅ Index
 
 [Index references](https://www.algolia.com/doc/api-reference/widgets/index-widget/js/)
 
 `Index` is the component that lets you apply widgets to a dedicated index. It’s useful if you want to build an interface that targets multiple indices.
 
-Not compatible as Meilisearch does not support federated search on multiple indexes.
-
-If you'd like to see federated search implemented please vote for it in the [roadmap](https://roadmap.meilisearch.com/c/74-multi-index-search?utm_medium=social&utm_source=portal_share).
+Using this component, instant-meilisearch does an http-request for each different Index added. More http requests are made when using the [`RefinementList`](#✅-refinementlist) widget.
```
Comment (mdubus) [inline README.md:348]: Do you want to keep the `Index` with the uppercase?

File: README.md [hunk 8]
```diff
@@ -672,6 +670,9 @@ The `refinementList` widget is one of the most common widgets you can find in a
 - ✅ templates: The templates to use for the widget.
 - ✅ cssClasses: The CSS classes to override.
 
+The `RefinmentList` widget uses the `disjunctive facet search` principle when using the `or` operator. For each different facet category used, an additional http call is made.
```
Comment (mdubus) [inline README.md:673]:
```suggestion
The `RefinementList` widget uses the `disjunctive facet search` principle when using the `or` operator. For each different facet category used, an additional http call is made.
```

File: README.md (hunk 7 above)
Comment (bidoubiwa) [reply README.md:348]: Yes since if the user uses the `Index` widget on the same index multiple times it still does that amount of HTTP requests. But you are right I'm going to improve that

File: README.md (hunk 8 above)
Comment (bidoubiwa) [reply README.md]: I dont like that word >:(

File: tests/multi-index-search.tests.ts (hunk 6 above)
Comment (bidoubiwa) [reply tests/multi-index-search.tests.ts]: I added the explaination. Instantsearch types telling me its read-only.

**Review by mdubus — DISMISSED** [review]
LGTM ✨🦕

File: tests/multi-index-search.tests.ts [hunk 9]
```diff
@@ -0,0 +1,270 @@
...
+      },
+    ])
+
+    const emptyMovies = emptyResponse.results[0]
+    const emptyGames = emptyResponse.results[1]
+
+    expect(emptyMovies.hits.length).toBe(0)
+    expect(emptyGames.hits.length).toBe(0)
+
+    const response = await customClient.search<Movies>([
```
[the remark this answers is not in the document]
Comment (bidoubiwa) [reply tests/multi-index-search.tests.ts:250]: Instantsearch requires a search method with a generic. Nonetheless, you are right that it makes no sense as there are multiple indexes now. 

I looked at the [code of algoliasearch](https://github.com/algolia/algoliasearch-client-javascript/blob/master/packages/client-search/src/methods/client/multipleQueries.ts) and it seems the issue is at their side as well.

File: cypress/integration/search-ui.spec.js [hunk 10]
```diff
@@ -38,24 +38,27 @@ describe(`${playground} playground test`, () => {
...
     const select = `.ais-SortBy-select`
-    cy.get(select).select('steam-video-games:recommendationCount:asc')
+    cy.get(select).select('games:recommendationCount:asc')
     cy.wait(1000)
     cy.get(HIT_ITEM_CLASS).eq(0).contains('Deathmatch Classic')
   })
 
   it('Sort by default relevancy', () => {
     const select = `.ais-SortBy-select`
-    cy.get(select).select('steam-video-games')
```
Comment (brunoocasali) [inline cypress/integration/search-ui.spec.js:48]: Big PRs usually hide many essential things reviewers don't get because they feel overwhelmed by the changes.

This change seems to be just a simple index name, a harmless change, but if you add ten more changes like that, the git diff is too complex to be helpful.

Btw, no need to move this to a different PR or something like that, it is just something to be in mind.

File: cypress/integration/search-ui.spec.js [hunk 11]
```diff
@@ -38,24 +38,27 @@ describe(`${playground} playground test`, () => {
...
-    cy.get(checkbox).eq(1).click()
+  it('click on facets ensure disjunctive facet search', () => {
+    const facet = `.ais-RefinementList-list`
+    const checkbox = `.ais-RefinementList-checkbox`
+    const facetCount = '.ais-RefinementList-count'
+    cy.get(facet).eq(0).find(checkbox).eq(1).click() // genres > action
+    cy.wait(1000)
+    cy.get(facet).eq(0).find(facetCount).eq(0).contains('5') // genres > action count
+    cy.get(facet).eq(1).find(facetCount).eq(0).contains('4') // players > multiplayer
     cy.wait(1000)
```
Comment (brunoocasali) [inline cypress/integration/search-ui.spec.js:61]: Is this last `wait` clause necessary?

File: cypress/integration/search-ui.spec.js [hunk 12]
```diff
@@ -78,6 +81,6 @@ describe(`${playground} playground test`, () => {
 
   it('Paginate Search', () => {
     cy.get('.ais-InfiniteHits-loadMore').click()
-    cy.get(HIT_ITEM_CLASS).should('have.length', 11)
+    cy.get(HIT_ITEM_CLASS).should('have.length', 12)
```
Comment (brunoocasali) [inline cypress/integration/search-ui.spec.js:84]: Why this has been changed?

File: scripts/jest_teardown.js [hunk 13]
```diff
@@ -0,0 +1,12 @@
...
+
+const HOST = 'http://localhost:7700'
+const API_KEY = 'masterKey'
+
+afterAll(async () => {
+  const client = new MeiliSearch({ host: HOST, apiKey: API_KEY })
+  await client.deleteIndex('movies')
+  const task = await client.deleteIndex('games')
+
+  await client.waitForTask(task.taskUid)
```
Comment (brunoocasali) [inline scripts/jest_teardown.js:11]: Don't you have a method that accepts more than one `uid`? (or I'm misleading because of other SDKs...)

Comment (bidoubiwa) [reply scripts/jest_teardown.js:11]: I don't have one in js :(

File: cypress/integration/search-ui.spec.js (hunk 12 above)
Comment (bidoubiwa) [reply cypress/integration/search-ui.spec.js:84]: because of the disjunctive facet search accepting more hits in return 😅

File: cypress/integration/search-ui.spec.js (hunk 10 above)
Comment (bidoubiwa) [reply cypress/integration/search-ui.spec.js:48]: I agree, thanks for pointing this out :)

**Review by brunoocasali — APPROVED** [review]

**Comment by bidoubiwa** [comment]
bors merge

**Comment by bidoubiwa** [comment]
bors merge
